Policy auth precedence fix

jrajahalme opened the pull request:
Note for review: This is blocked by #26344 and needs to be rebased after it merges. First two commits are from this.

Explicitly specified authentication mode should apply to all traffic that matches the given policy map key. For example, if an L3-only policy specified that authentication is required, then it should be required also on all L3/L4-policies on the same L3, even if not specified in the L3/L4-policy. Similarly, an L3-only auth policy should apply to to all ports on the given L3, even if there is an L4-only policy that has no authentication requirement. In the most general case, an Allow-all policy with authentication requirement should apply to all traffic, unless authentication is explicitly disabled for specific L3, L4, or both.

To achieve this, we must scan all the policy map keys, when adding a new entry, as the new entry's auth type may have to be overridden by the explicit auth type of a more general entry, or the auth type of more specific entries may have to be updated due to the explicit auth type on the new entry. This is somewhat similar to how deny rules take precedence over more specific entries.

To limit the additional computational cost of this map scanning, we track the use of authentication (and deny rules) on the ingress/egress policy level, and skip the scanning that is not necessary when the policy does not contain any authentication (or deny) rules.

The commit to merge DerivedFromRules (from PR #26346) is included here to allow for the unit tests in the last commit to reliably track the DerivedFromRules labels of the entries added by `toMapState()`.

jrajahalme committed:
policy: Streamline ToMapState() use

Users of ToMapState() combine the results from a single-filter ToMapState
call to a policy-wide MapState using DenyPreferredInsert, which causes
the heavy logic needed for giving preference to deny entries to be
computed at least twice per entry, once within ToMapState, and then again
in the caller. Since DenyPreferredInsert may add new entries, the 2nd
round may actually involve more entries than the first.

Change the ToMapState to match its current use by:
- take the final map where entries are to be added as a parameter to
  avoid the need for an intermediate return MapState
- take in a filter function so that the caller may update an entry
  without iterating over the mapstate again (e.g., replace temporary proxy
  port with the actual proxy port), and the caller has the option to skip
  adding the entry by returning 'false'.
- take Keys arguments for 'adds' and 'deletes' that are passed to
  DenyPreferredInsertWithChanges().

This change reduces the computational cost of policy updates, and makes
it easier to reason about the MapState processed with
DenyPreferredInsertWithChanges, now that it is done in a single pass,
rather than multiple passes like before.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add L4DirectionPolicy

Add L4DirectionPolicy containting the L4PolicyMap, so that additional
fields can be added in later commits.

This should be a pure refactor without functional changes.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add UpdateRedirects()

Add EndpointPolicy.UpdateRedirects() to keep more of the policy internals
within the policy package.

Rename GetLabels to GetLabelsLocked and note in the implementations that
the lock of the implementing object is held when it is called.

Make ToMapState non-exported, and simplify callers now that they are all
in the policy package.

Rename computeDesiredL4PolicyMapEntries and
computeDirectionL4PolicyMapEntries to 'toMapState' for their respective
receivers, as they do the same function and eventually call the
L4Filter.toMapState.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Track policy features

Track policy features in use across all PortRules in a given traffic
direction, and skip policy computations not needed when specific features
are not in use.

This is needed as processing deny rules involves scanning all MapState
entries for each new entry (O(N^2)), which can be skipped when there are
no deny rules in the policy. This will also be used for Authentication
policy processing for the same reason, so that if Authentication is not
used, there is no additional CPU overhead.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpoint: Do not override deny entries with proxy redirects

Use DenyPreferredInsert instead of directly manipulating policy map state
to make sure deny entries are not overridden by new proxy redirect
entries.

Prior to this fix it was possible for a proxy redirect to be pushed onto
the policy map when it should have been overridden by a deny at least in
these cases:

- L3-only deny with L3/L4 redirect: No redirect should be added as the L3
  is denied

- L3-only deny with L4-only redirect: L4-only redirect should be added
  and an L3/L4 deny should also be added, but the L3/L4 deny is only added
  by deny preferred insert, and is missed when the map is manipulated
  directly. A new test case verifies this.

It is clear that in the latter case the addition of the redirect can not
be completely blocked, so we can't fix this by making AllowsL4 more
restrictive. But also in the former case it is possible that the deny
rule only covers a subset of security identities, while the redirect rule
covers some of the same security identities, but also some more that
should not be blocked. Hence the correct fix here is to leave AllowsL4 to
be L3-independent, and cover these cases with deny preferred insert
instead of adding redirect entries to the map directly.

This commit also contains a related change that allows a redirect entry
to be updated, maybe with a changed proxy port. I've not seen evidence
that this is currently fixing a bug, but it feels like a real
possibility.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Export DenyPreferredInsertWithChanges, make revertible

Export DenyPreferredInsertWithChanges and make it revertible by taking a
map of old values as a new optional argument.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joestringer commented on pkg/policy/l4.go:
```diff
@@ -479,21 +497,25 @@ func (l4 *L4Filter) GetListener() string {
-//	Entry with any other port #: Proxy redirection is required for this key,
-//	                             caller must replace the ProxyPort with the actual
-//	                             listening port number.
+// toMapState converts a single filter into a MapState entries added to 'mapState'.  'filter' is
```
nit: I think that the "filter" parlance here is a little bit misleading. Probably the reader can ascertain that it's not really a pure filter given that the value is passed by reference, which implies that there's no bets about whether its content is modified, but that's probably all the more reason to define clearer semantics for it. Initially I thought Mutator, but actually a mutator doesn't typically return a result. So is it a `FilterMutator`?

Incidentally, giving this a named type like you did with the `getProxyPortFunc` would also provide a better space to document the expected semantics for the function.

jrajahalme replied:
Added a function type for this:
```
// 'mapStateEntryCallback' is a function called for each entry before adding to a MapState. If the
// function returns 'true', the entry is added, otherwise not. The function gets a reference to the
// entry so that it may update it's value (e.g., the proxy port).
type mapStateEntryCallback func(key Key, value *MapStateEntry) bool
```

joestringer commented on pkg/policy/resolve.go:
```diff
@@ -157,40 +157,91 @@ func (p *EndpointPolicy) Detach() {
+
+type getProxyPortFunc func(*L4Filter) (proxyPort uint16, ok bool)
+
+// UpdateRedirects updates redirects in the policy by calling the given 'filter' to obtain a proxy port number to use.
```
nit, getting the proxy port isn't a filter by any meaning of `filter` I'm aware of: 
```suggestion
// UpdateRedirects updates redirects in the policy using the provided function to obtain a proxy port number to use.
```

joestringer replied:
Also now that there's `adds` and `updated`, these should probably also be documented. It's becoming less clear to me whether this function modifies the received object or if it's supposed to be a function that doesn't change `p` but just uses info from `p` to populate `adds`/`updated`.

jrajahalme replied:
Changed to:
```
// UpdateRedirects updates redirects in the EndpointPolicy's PolicyMapState by using the provided
// function to obtain a proxy port number to use. Changes to 'p.PolicyMapState' are collected in
// 'adds' and 'updated' so that they can be reverted when needed.
func (p *EndpointPolicy) UpdateRedirects(ingress bool, identities Identities, getProxyPort getProxyPortFunc, adds Keys, updated MapState) {
```

joestringer commented on pkg/policy/resolve.go:
```diff
@@ -157,40 +157,91 @@ func (p *EndpointPolicy) Detach() {
+	// Nothing to do if L4Policy does not exist
+	if p.L4Policy == nil {
+		return
+	}
```
nit: At face value I'm a little confused why a caller would update the redirects if there's no policy to update, is this defensive coding or is there a known case where this could be nil?

jrajahalme replied:
This pointer was initially `nil` when a policy had not yet been computed. I added a refactor commit that inlines `selectorPolicy.L4Policy` so that these pointer checks are never needed. An initialised but empty `L4Policy` can safely be used.

joestringer commented on pkg/policy/l4.go:
Side note, I think we have a meta question to ponder as a project, which is the degree to which we lean on anonymous functions. In limited circumstances I find them quite useful. At the same time I am a little worried that if we add them too widely it'll become increasingly difficult to `grep` for and trace when reading the code later on.

joestringer replied:
I think that in the usage in this PR it's fairly straightforward: One case is a filter+mutator which seems like a pretty well known pattern for anonymous functions like this, and it's comparatively easy to trace since the caller provides the function. The proxy port fetcher is similarly relatively well constrained.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -479,21 +497,25 @@ func (l4 *L4Filter) GetListener() string {
+// Keys of any added or deleted entries are added to 'adds' or 'deletes', respectively, if not nil.
 // PolicyOwner (aka Endpoint) is locked during this call.
-func (l4Filter *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirection.TrafficDirection, identities Identities) MapState {
+func (l4Filter *L4Filter) toMapState(p *EndpointPolicy, features policyFeatures, filter func(key Key, value *MapStateEntry) bool, adds, deletes Keys) {
```
nit: As much as I am scared to ask for further refactoring of these functions, I feel like this "to map state" is getting a bit out of control in its parameters.

Before looking into the details, I can see that it operates on an L4 filter, then it also takes an endpoint policy which might be an input or output and may or may not have some relationship with the l4filter, then there's a set of features that impact the map state generation, a filter to decide whether to create mapstate entries, and then maybe two other output parameters in the form of adds or deletes, which implies it's not just statically converting the L4filter into a single corresponding mapstate, but actually calculating a diff against something... How does someone reason about the impact on the data being passed around here? Ideally I'd prefer to see this broken into smaller functional steps that each operate in a clearly defined manner with respect to one another, ideally with clear distinctions of mutable vs. immutable state.

Given that this change is complex enough as it is, it probably doesn't make sense to further refactor this code in this PR, but I'm just going to note here that I feel like the already-complicated policy logic is getting even more difficult to understand.

joestringer replied:
Coming back to this: the first parameter is now the way that the named ports selection gets dependency injected *and* it's one of the output parameters. Then the `filter` is also a dependency-injected function, so between `GetNamedPortLocked()` and `filter()` there's two different patterns being used here for providing implementations to this function to partially resolve the logic internally.

If this is actually modifying `EndpointPolicy`, then that makes me wonder if the receiver should be `EndpointPolicy` in this case. :thinking: not sure what the best practices are for that sort of case. But even then it's a bit odd given that there's three output parameters: `EndpointPolicy.PolicyMapState`, `adds`, and `deletes`. I guess as long as we clearly define the inputs and outputs and describe the way that this function uses the different mutator functions along with the input in order to generate the multiple outputs, then probably this is fine.

joestringer replied:
`filter.toMapState(...)` does generally sound like it takes the filter input and injects it into the outputs, though the "return the output" is a little bit more obvious than passing output parameters in. At the same time, given that this is doing partial calculation I can also understand why it makes sense for the output parameters to be passed in.

Even though my initial feedback above was that it feels like there's too many parameters already, having reviewed the PR I'm wondering if it should even go further, something like:

```
func (l4Filter *L4Filter) toMapState(namedPortResolverFunc, filterMutatorFunc, mapState, adds, deletes Keys) {
```

(for extra points I guess we might at some point try to bound the three output parameters together, but perhaps there's good reasons to keep those separate based on how they're stored & used -- I'd guess that `mapState` becomes the long-lived store of the Desired policy map state, whereas adds/deletes are the intermediate diffs that are useful for efficiently manipulating kernel map state. So maybe then it's just the adds and deletes that get combined into a `policyMapChanges` structure)

joestringer replied:
Note that at the point of providing the above feedback, I haven't walked up and down the stack to understand the new `toMapState` functions at different layers so it's possible that I'm oversimplifying while trying to understand one layer of the policy engine in detail. Food for thought anyway, and this whole thread is more for discussion and mulling over thoughts on making the policy engine easier to read and engage with... maybe I'm just rabbitholing. Moving on now.

joestringer replied:
Ah I also neglected to mention the policyFeatures output parameter as well.

jrajahalme replied:
This get's even a bit more complicated when #26344 gets merged. I'll see if I can rein in the function parameters, but I'll need to rebase on top of #26344 to make more sense out of this.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -479,21 +497,25 @@ func (l4 *L4Filter) GetListener() string {
-//	                             caller must replace the ProxyPort with the actual
-//	                             listening port number.
+// toMapState converts a single filter into a MapState entries added to 'mapState'.  'filter' is
+// called for each entry before adding to 'mapState', if filter returns 'true', the entry is added,
```
There is no parameter here `mapState`. Please update the comment for this function to clarify which parameters are inputs, which are outputs, how the function processes the inputs, and the API around the filter/mutator.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -172,30 +169,6 @@ func (e *Endpoint) writeHeaderfile(prefix string) error {
-// implementation uses the identity allocator. This means that the locking
-// patterns from this code path will differ from the other policy calculation
-// cases!
-func (p *policyIdentitiesLabelLookup) GetLabels(id identity.NumericIdentity) labels.LabelArray {
```
This commit adds a SelectorCache mutex lock to one of the policy generation hotpaths. How do we know that's safe? We've been dealing with poor lock ordering constraints between ipcache, endpoint and policy for almost two years now with the latest round of problems being actively solved right now. I'm pretty paranoid about any locking pattern changes in this area at this point.

At minimum I would like to see a detailed explanation of why we think this is safe. I'd probably prefer to see it as a separate commit by its own, probably at the end, so we can revert it if it turns out to introduce another deadlock. And if it's not crucial to the solution, maybe even just defer making this specific change to a later PR with more comprehensive reasoning through the overall policy generation flow to provide stronger guarantees about lack of deadlock.

jrajahalme replied:
Will investigate and revert if needed.

jrajahalme replied:
Yes, I think I need to revert this and not require this new selector cache locking for the reasons you stated. I'll add a comment to the GetLabels here to make it easier to understand that this exists for the sole purpose of not being able to lock selector cache while endpoint mutex is held!

joestringer commented on pkg/policy/l4.go:
```diff
@@ -1019,9 +1050,44 @@ func addL4Filter(policyCtx PolicyContext,
+type L4DirectionPolicy struct {
+	PortRules L4PolicyMap
+
+	// features tracks properties of PortRules to skip code when features are not used
+	features policyFeatures
```
nit: I was under the impression that `L4DirectionPolicy` was an immutable definition of desired state. But the `policyFeatures` now seems like a mutable cache that gets updated when the filters in the policy get updated. I can't tell if this is a problem yet or not, but it feels like it could cause consistency issues at some point down the line. I don't have a counterproposal.

joestringer replied:
Maybe one way to ease my concern here would be to process the need for deny / authentication during initial policy processing and creation of the L4DirectionPolicy instead of during `attach()`.

jrajahalme replied:
`policyFeatures` are immutable everywhere after attach. We compute new filters for each new policy update, they are only ever updated during the policy update, when they need to be merged. So attach() marks the end of the initial policy processing. After this the `selectorPolicy` is immutable, gets shared for multiple endpoint policies that then make use of the (now immutable) `policyFeatures`. Or this is the intent, anyway!

joestringer replied:
Ah, somehow I figured that `attach()` happens potentially multiple times and hence could cause the state to become unexpectedly mutable. But actually that may be a misunderstanding on my side.

joestringer replied:
```
$ git grep "Attach("
pkg/policy/l4.go:func (l4 *L4Policy) Attach(ctx PolicyContext) {
pkg/policy/l4_test.go:  policy.Attach(testPolicyContext)
pkg/policy/repository.go:       calculatedPolicy.Attach(&policyCtx)
pkg/policy/resolve.go:func (p *selectorPolicy) Attach(ctx PolicyContext) {
pkg/policy/resolve.go:          p.L4Policy.Attach(ctx)
```

Seems like once per identity and each identity has a unique policy. Then as you state, selectorPolicy gets recalculated and atomically set with the expectation of immutability. So we should be good on this front?

jrajahalme replied:
Yes :-)

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -242,7 +247,7 @@ func (keys MapState) RemoveDependent(owner Key, dependent Key) {
 }
 
-// MergeReferences adds owners and dependents from entry 'entry' to 'e'. 'entry' is not modified.
+// MergeReferences adds owners, dependents, and DerivedFromRules from entry 'entry' to 'e'. 'entry' is not modified.
```
how does this merging of the labels impact production environments? I'm thinking from two perspectives:
1. Debuggability. This feels like it might potentially improve the ability to debug map state changes, but I'm not quite sure how a developer or user might do that.
2. Performance. This seems like a whole bunch of extra string manipulation and potentially allocations. If this is called often, it could impact policy generation times.

jrajahalme replied:
1. Logging now contains this, also there is `GetRealizedPolicyRuleLabelsForKey()`, but that seems to be used for unit testing only, currently?

2. Addressed this by adding a new commit that exports `labels.MergeSorted()` that now merges in place and minimises memory allocations.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -729,15 +927,15 @@ func (keys MapState) DetermineAllowLocalhostIngress() {
 // AllowAllIdentities translates all identities in selectorCache to their
 // corresponding Keys in the specified direction (ingress, egress) which allows
-// all at L3, without requiring authentication.
+// all at L3, WITHOUT REQUIRING AUTHENTICATION.
```
nit: was this capitalization deliberate?

jrajahalme replied:
Yes, but will revert if you think it is too much of an attention grab. I had to reason about the auth requirements on every call site and having this pop out makes it more obvious why the combo `ExplicitAuthType, AuthTypeDisabled` is used.

joestringer replied:
It certainly grabbed my attention :sweat_smile: 

What wasn't clear to me is the implications here. I'd prefer a more verbose explanation, even with scary words like "Warning", or "Be aware" etc. that help the reader to reason about this rather than just seeing the capital letters and being left to figure that part out :-)

jrajahalme replied:
Changed:
```
-// all at L3, WITHOUT REQUIRING AUTHENTICATION.
+// all at L3.
+// Note that this is used when policy is not enforced, so authentication is explicitly not required.
```

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -498,8 +526,178 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+
+// authPreferredInsert applies AuthType of a more generic entry to more specific entries, if not
+// explicitly specified.
+func (keys MapState) authPreferredInsert(newKey Key, newEntry MapStateEntry, features policyFeatures, adds, deletes Keys) {
```
This function and the logic in `IsSuperSetOf()` seems awfully similar to equivalent logic in deny policy logic handling. On one hand, having separate implementations seems like potentially a good way to prevent auth policies from causing a regression in deny policies. On the other hand, I'm concerned that now we may end up with two entirely different implementations of conflict resolution for map entries.

Is there a way to share the core implementation with deny preferred inserts here? I'd imagine there are maybe some subtle differences in the implementations, but I feel like supporting one giant set of confusing conflict resolution statements that generically applies to both cases would be easier than maintaining two entirely different sets.

joestringer replied:
I've reached my policy confusion quota for the day, so I've stopped before reviewing any of this prefered insert logic.

jrajahalme replied:
Keeping them separate makes it obvious that auth policies can not revert any deny rules with non-deny rules. The logic is also subtly different and I'd rather get it right first and optimise by combining the loops later.

joestringer replied:
Ack, reasonable tradeoff. Is there a TL;DR/explanation on the subtly different logic?

jrajahalme replied:
For one the super/subsetting is different when auth does not need to deal with CIDRs. Then there is the fact that authentication requirement can be (explicitly) disabled on a more specific rule, whereas a more generic deny can not be overridden by an allow rule. I think could probably refactor this in future for efficiency by combining the loops (so looping through the mapstate entries only once even when both deny and auth features are in use) with the cost of making it harder to understand. Complexity would increase from the need to collect more state to the end of the loop to find out if later iterated deny entries will override any of the (added) auth entries.

joestringer reviewed (changes requested):
My biggest concern in the PR so far was the change in locking behaviour for identity labels lookup which adds a new selectorcache lock usage in one of the paths. See the feedback below on that.

The next biggest was around the new function in the mapstate package for evaluating the merging of policymap state entries for auth rule cases, where it wasn't clear to me whether we considered sharing the implementation with the deny prefered insert cases. Logically they seem similar to me, but I may be missing some context around why that choice makes sense.

I stopped short of reviewing the merging logic in mapstate, and I didn't get to the distillery tests either.

The rest of my comments are mostly minor nits that can be ignored, bear with my mild complaints about the complexity of the policy code :-) It never really gets easier to grok this logic.

jrajahalme committed:
labels: Export MergeSorted, sort in place

Export MergeSorted and make it more efficient by adding entries to the
receiver as needed.

Use MergeSorted in policy where it is known that the LabelArrayLists are
in sorted order already.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Merge DerivedFromRules when merging map state entries

Keep track of all contributing rule's labels in merged map state entries.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Inline selectorPolicy.L4Policy

There is no reason selectorPolicy.L4Policy needs to be a pointer as we
can initialize an empty L4Policy for it.

Refactor by inlining L4Policy instead of pointing to it, as this
simplifies code by removing the need for pointer nil checks.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

nebril reviewed (commented):
LGTM to me overall, but as I am not that familiar with policy code, leaving a comment instead of Approve.

jrajahalme committed:
policy: Handle l4-only covering l3l4 rules explicitly

When considering if an l3l4 rule can be skipped due to a l4-only
(wildcard L3) rule on the same filter we must also consider the auth
types of both rules. More specifically, the l3l4 rule must be poshed to
the datapath if it has an authentication policy while the l4-only rule
does not, or if both have different auth policies. This opens up a
possiblity that due to this auth type discrepency the redirect status of
the l4-only rule must be explicitly applied to the l3l4 rule as the l3l4
rule can no longer be skipped in all cases.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Apply authentication from more general rules to more specific ones

Apply authentication policy from more general rules to more specific
rules, unless explicitly specified.

Restructure addKeyWithChanges to make it easier to understand:

- if an old entry exists, the new one is merged to it, honoring the
precedence rules for deny, proxy redirection, and auth type.

- new entries get their own containers for owners, dependents, and
DerivedFromRules so that each entry in MapState has separate
containers. This was already the case previously, but the code was had to
reason about.

redirectPreferredInsertWithChanges() is no longer needed as a separate
function, as addKeyWithChanges now follows the same logic, but instead of
simply overriding the old entry with the new one considers the precedence
rules for deny, proxy redirect and auth type.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: return hasAuthType from GetAuthType

Make GetAuthType return both if the auth type is explicitly present, and
the auth type itself.

This becomes necessary when considering if the auth type of a more
generic entiry should override the one in a more specific entry in
future.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Refactor toMapState arguments

Simplify code by grouping change tracking arguments to a single "changes"
argument.

Change location of "identities" argument in mapstate functions to be more
consistent with 'toMapState'.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Use addKeysWithChanges in AddvisibilityKeys

Use addKeysWithChanges in AddVisibilityKeys instead of directly
manipulating the maps to keep track of changes in a uniform fashion. Also
record changes when adding dependents to existing keys.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpoint: Use RevertChanges for visibility redirects too

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Simplify logic in addKeyWithchanges()

Restructure addKeyWithChanges to make it easier to understand:

- if an old entry exists, the new one is merged to it, honoring the
precedence rules for deny, proxy redirection, and auth type.

- new entries get their own containers for owners, dependents, and
DerivedFromRules so that each entry in MapState has separate
containers. This was already the case previously, but the code was had to
reason about.

- when storing old values for reverting, clone containers so any futher
changes in the values in the map will be change the stored old values.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Fix PerSelectorPolicy.Equal

The receiver of an autogenerated DeepEqual function can not be nil, must
check for it explicitly.

Generated api.Authentication.DeepEqual returns false whenever 'other' is
nil, even if the receiver is also nil.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -255,6 +310,11 @@ func (e *MapStateEntry) MergeReferences(entry *MapStateEntry) {
+
+	// merge DerivedFromRules
+	if len(entry.DerivedFromRules) > 0 {
+		e.DerivedFromRules.MergeSorted(entry.DerivedFromRules)
```
I recall my final piece of feedback on the parent PR around potential reference leaks when we go through these steps and then need to revert. Currently in this commit I see this new merge + sort added but I don't see this merge/sort getting unwinded upon revert. Is that a problem?

jrajahalme replied:
This merge should be a no-op when the entries are deeply equal, and the old value was stored for reverting by the calling function if the entries are not deeply equal.  It would be clearer to make this explicit by not calling into this merge code at all in the case of the entires being deeply equal. I'll add a commit to that effect.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -26,69 +26,39 @@ import (
+// covers returns true if 'l4rule' has the effect needed for the 'l3l4rule', when 'l4rule' is added
+// to the datapath, due to the l4-only rule matching if l3l4-rule is not present. This determination
+// can be done here only when both rules have the same port number (or both have a wildcarded port).
+func (l4rule *PerSelectorPolicy) covers(l3l4rule *PerSelectorPolicy) bool {
```
:heart: I was initially reticent to drop the truth table, since once you get your head around it, you can directly compare the various cases.. but with the way that the new outer function reasons through the precedence I think this new version is easier to understand.

I'm not exactly sure whether `covers()` is equivalent/opposite if you swap the receiver / parameter, it seems like this is assuming that the wildcard rule must be the receiver and the more specific rule is the parameter. But given that assumption, the logic looks correct. And as far as I can tell, that's the only way that this function is used anyway :+1: 

(no changes requested here, LGTM)

jrajahalme replied:
Yes, the naming of the receiver and the argument is meant to convey the intended call pattern.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+//
+// Specificity score for all possible superset wildcard patterns. Datapath requires proto to be specified if port is specified.
+// x. L3/proto/port
+//  1. */*/*
+//  2. */proto/*
+//  3. */proto/port
+//  4. ID/*/*
+//  5. ID/proto/*
+//     ( ID/proto/port can not be superset of anything )
```
Upon first read, I was expecting this to be either the same order as the policy match precedence in `bpf/lib/policy.h`, or the inverse order. But the actual order here is mostly the inverse of the datapath, but with (4) having a higher priority here than in the datapath.

Maybe I'll come back to this, I'm not sure what the implications are here or whether this needs to directly map to the datapath precedence.

joestringer replied:
Given that this function is determining superset membership strictly, hence there is no superset relationship between an L3-only entry and an L4-only entry, I've concluded that this specificity ordering isn't consequential in its difference from the datapath. Those cases would return 0, and the calling code then decides to synthesize more specific L3+L4 entries to enforce the policy auth semantics that the policy requires.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+// same key. Hence, a L3-only key is not a superset of a L4-only key, as the L3-only key would match
+// the traffic for the given L3 only, while the L4-only key matches traffic on the given port for
+// all the L3's.
+// Returns 0 if the receiver key is not a superset of the argument key.
```
nit: With a name like `IsSuperSetOf()`, I would have expected either:
1. The result to be a boolean or at least have one boolean result
2. With an int result, the bigger the return value, the "more of a superset" it is.

That said, while I think it's taking a bit for me to get my head around these semantics, this may end up making the caller's code more straightforward to reason about, so I don't necessarily have any immediate feedback to improve it.

joestringer replied:
I guess... summary: Naming is hard ;-)

I did sorta like the `covers()` naming used in previous versions, but that doesn't really say anything about specificity either. The specificity result makes this harder to name, but that's also a core semantic to the function usage. So yeah no specific solutions from my side. LGTM.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+					// only override the AuthType for the L3 & L4 combination,
+					// not L4 in general.
+					//
+					// These need to be collected and added after it is known if
```
nit, I think this might improve clarity? Feel free to take or skip:

```suggestion
					// These need to be collected and only added if
```

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+			}
+		} else {
+			// New entry has an explicit auth type.
+			// Check it the new entry is the most specific superset of any other entry with the default auth type,
```
nit:
```suggestion
			// Check if the new entry is the most specific superset of any other entry with the default auth type,
```

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+	return 0
+}
+
+// authPreferredInsert applies AuthType of a more generic entry to more specific entries, if not
+// explicitly specified.
```
nit:

```suggestion
// authPreferredInsert applies AuthType of a more generic entry to more specific entries, if not
// explicitly specified.
//
// This function is expected to be called for a map insertion after deny
// entry evaluation. If there is a map entry that is a superset of 'newKey'
// which denies traffic matching 'newKey', then this function should not be called.
```

joestringer replied:
(gathered from the `IsDeny` logic below and the caller usage, but it wouldn't hurt to make this a bit more explicit, I guess?)

joestringer reviewed (approved):
Mostly minor nits, just one main open question at the top of my feedback below. :+1:

nathanjsweet commented on pkg/policy/mapstate.go:
```diff
@@ -388,11 +502,29 @@ func protocolsMatch(a, b Key) bool {
-func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStateEntry, adds, deletes Keys, identities Identities) {
+func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStateEntry, identities Identities, features policyFeatures, changes ChangeState) {
+	// Skip deny rules processing if the policy in this direction has no deny rules
+	if !features.contains(denyRules) {
```
I'm pretty sure this is wrong, or at least not specific enough. In `EndpointPolicy.ConsumeMapChanges` features is set as a join of the "Ingress.features" and "Egress.features" of an L4Policy. What if the `newKey` and `newEntry` are derived from the Ingress of the policy which doesn't contain any deny rules, but the Egress does? If I'm reading this correctly then we will incorrectly skip calling `authPreferredInsert`.

jrajahalme replied:
This skip of the deny logic when no denies are present is purely an optimization made possible by tracking of policy features that was added mainly to cut the additional cost of the auth-logic when authentication is not used in the policy. My recollection is that on the the ConsumeMapChanges code path the direction specific features would have been harder to use, as the collected map changes can contain entries in both directions. In that case we end up doing some unnecessary deny/auth logic for the direction where those features were not actually present.

`authPreferredInsert` call is not skipped in any case, but if deny feature is indicated it will be done at the bottom of `denyPreferredInsert` after scanning through all the mapstate entries.

nathanjsweet commented on pkg/policy/mapstate.go:
```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+		if newEntry.hasAuthType == DefaultAuthType {
+			// New entry has a default auth type.
+			// Fill in the AuthType from more generic entries with an explicit auth type
+			max_specificity := 0
```
nit: Any reason this shouldn't be camelCase?

jrajahalme replied:
No reason, thanks for noticing, will fix!

nathanjsweet commented on pkg/policy/mapstate.go:
```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+					continue
+				}
+
+				// Find out if 'k' is a more specific superset of 'newKey' than seen so far
```
```suggestion
				// Find out if 'k' is an identity-port-proto superset of 'newKey'

nathanjsweet commented on pkg/policy/mapstate.go:
```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+	if k.TrafficDirection != other.TrafficDirection {
+		return 0 // TrafficDirection must match for 'k' to be a superset of 'other'
+	}
+	if k.Identity == 0 {
```
Do we need to think about CIDRS in this method?
```suggestion
	if k.Identity == 0 ||
		ip.NetsContainsAny(keyEntry.getNets(identities, key.Identity),
			otherEntry.getNets(identities, other.Identity)) {
```
Obviously the method would need to be refactored to accept the entries, and also the `identities` interface. I'm fairly certain the specificity math does not need to change even if this gets added.

joestringer replied:
My gut feeling is that CIDRs don't need to be considered, because prior to policy evaluation the ipcache will already apply LPM to identify the most specific identity for the policymap entry. Given that this identity is label-selected in the same way as any other policy evaluation logic, for more specific CIDRs there will be additional identities and hence additional map entries to cover each superset or subset CIDR. The only conflicts we need to deal with here are interactions between wildcard entries (eg "applies to all L3s" or "applies to all L4s") and more specific identity policies (where some of the L3/L4 fields are the same, while the others are wildcarded). 

Put another way, CIDR identity ambiguity is already handled by the LPM ipcache lookup earlier in the datapath, so this means that the policy engine doesn't have to account for it; the policy posture for each explicit CIDR is individually evaluated.

That said, if you have a particular set of overlapping policy rules in mind to consider for this case, that would be super helpful both to reason through and also to ensure the tests are covering that case.

nathanjsweet replied:
Joe and I discussed this in person and decided that CIDRs aren't relevant to Auth. That is to say, no CIDR policy will currently select an endpoint that Cilium can enforce authentication or authorization on.

joestringer replied:
Related, we dug fairly deep into some of the reasons why the deny policy resolution logic needs to take into account CIDR identity supersets, and I've tried to document some of the background in this PR: https://github.com/cilium/cilium/pull/26720 . This may also be a helpful reference to contrast with the auth policy cases to see how CIDR identity supersets may or may not impact this policy evaluation logic in future.

nathanjsweet reviewed (changes requested):
There are some improvements to be made in mapstate.go. `denyPreferredInsert` and `authPreferredInsert` are very similar, I wonder if there is a way that we can generalize them together?

jrajahalme committed:
policy: Do not merge mapstate entries when they are deeply equal

Merging of deeply equal map state entries is not necessary, and it is
clearer to skip the merge in this case.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Editorial corrections

Comment corrections for readability and camel-casing a local variable.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
